Snowflake.Data/Core/SFResultSet.cs

```diff
@@ -2,91 +2,83 @@
  * Copyright (c) 2012-2017 Snowflake Computing Inc. All rights reserved.
  */
 
+using System.Threading.Tasks;
 using Common.Logging;
 using Snowflake.Data.Client;
 
 namespace Snowflake.Data.Core
 {
     class SFResultSet : SFBaseResultSet
     {
-        private static readonly ILog logger = LogManager.GetLogger<SFResultSet>();
+        private static readonly ILog Logger = LogManager.GetLogger<SFResultSet>();
+        
+        private int _currentChunkRowIdx;
 
-        private QueryExecResponseData execFirstChunkData;
+        private int _currentChunkRowCount;
 
-        private int currentChunkRowIdx;
+        private readonly int _totalChunkCount;
+        
+        private readonly SFChunkDownloader _chunkDownloader;
 
-        private int currentChunkRowCount;
-
-        private int totalChunkCount;
-
-        private int currentChunkIndex;
-
-        private SFChunkDownloader chunkDownloader;
-
-        private SFResultChunk currentChunk;
+        private SFResultChunk _currentChunk;
 
         public SFResultSet(QueryExecResponseData responseData, SFStatement sfStatement)
         {
-            execFirstChunkData = responseData;
             columnCount = responseData.rowType.Count;
-            currentChunkRowIdx = -1;
-            currentChunkRowCount = responseData.rowSet.GetLength(0);
-            currentChunkIndex = 0;
-
+            _currentChunkRowIdx = -1;
+            _currentChunkRowCount = responseData.rowSet.GetLength(0);
+           
             this.sfStatement = sfStatement;
 
             if (responseData.chunks != null)
             {
                 // counting the first chunk
-                totalChunkCount = responseData.chunks.Count + 1;
-                chunkDownloader = new SFChunkDownloader(responseData.rowType.Count,
+                _totalChunkCount = responseData.chunks.Count + 1;
+                _chunkDownloader = new SFChunkDownloader(columnCount,
                                                         responseData.chunks,
                                                         responseData.qrmk,
                                                         responseData.chunkHeaders);
             }
 
-            currentChunk = new SFResultChunk(responseData.rowSet);
+            _currentChunk = new SFResultChunk(responseData.rowSet);
+            responseData.rowSet = null;
 
             sfResultSetMetaData = new SFResultSetMetaData(responseData);
 
             updateSessionStatus(responseData);
             isClosed = false;
         }
 
-        internal override bool next()
+        internal override Task<bool> NextAsync()
+        {
+            return Task.FromResult(Next());

```

Comment:
So I am assuming that this is synchronous operation, right? In most cases I guess synchronous Next() is okay. 

However, if the next line is start of a new chunk and download from s3 has not finished yet, it is possible that this consumer thread should be blocked so that waiting download being finished.

Reply:
It is possible that this call may be synchronous under those situations. However to make it not synchronous by wrapping Next in a Task.Run() results in so much overhead that it causes performance problems.

We would need to duplicate (or decompose) the code between Next and NextAsync and make sure that only the call to _chunkDownloader.GetNextChunk is wrapped in a task (or made asynchronously)